# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [5]:
# Подсчитаем процент строк с пропусками относительно всего датафрейма
display("{:.2%}".format(data['days_employed'].isna().sum() / len(data['days_employed'])))

'10.10%'

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
# Сгруппируем наши данные, что бы можно было подсчитать как количество детей влияет на количество должников
children_debt = data.groupby('children').agg({'debt' : ['mean', 'count', 'sum']})

In [24]:
children_debt.round(3)

debt             
           mean  count   sum
children                    
0         0.075  14091  1063
1         0.092   4808   444
2         0.095   2052   194
3         0.082    330    27
4         0.098     41     4
5         0.000      9     0

**Вывод:** 

   Из этих данных мы видим, что среди заемщиков без детей меньше всего должников всего 7,5% из общего количества в 14091 клиента. Среди заемщиков с 1 ребенком 9,2% должников с количесвом клиентов в 4808 человек, среди заемщиков с 2 детьми 9,5% должников с количеством клиентов в 2052 человека.
    Колчество клиентов с 3-мя, 4-мя и 5-ю детьми невелико 330, 41 и 9 человек соответственно, поэтому хоть процент должников в этих группах и соответсвует полученным данным, тяжело назвать эту выборку реперезентативной.
    
   Из этих данных мы можем сделать вывод, что чем выше количество детей у заемщика, тем выше количество должников.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
### Сгруппируем наши данные, что бы можно было подсчитать как семейный статус влияет на количество должников.

family_debt = data.groupby('family_status_id').agg({'debt' : ['mean', 'sum', 'count']})

In [26]:
family_debt.round(3)

debt            
                   mean  sum  count
family_status_id                   
0                 0.076  927  12261
1                 0.093  385   4134
2                 0.066   63    951
3                 0.071   84   1189
4                 0.098  273   2796

**Вывод:** 
Из этих данных мы видим, то наименьшее количество долгов у людей ранее состоявших в официальных отношениях - потерявшие своего партнера и разведенные примерно 6,6% и 7,1% соответственно. Далее идет группа людей состоящих в официальных отношений 7,5%. И самые "ненадежные" клиенты банка в гражданском браке и не-женатые/замужние с 9,3% и 9,8% соответственно.

Отсюда можно сделать вывод, что люди состоящие в официальных отношениях более надежные заемщики.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
### Сгруппируем наши данные, что бы можно было подсчитать как величина дохода влияет на количество должников.

income_debt = data.groupby('total_income_category').agg({'debt' : ['mean', 'sum', 'count']})

In [28]:
income_debt.round(3)

debt             
                        mean   sum  count
total_income_category                    
A                      0.080     2     25
B                      0.071   354   5014
C                      0.085  1353  15921
D                      0.060    21    349
E                      0.091     2     22

**Вывод:** 
Из этих данных видно, что заемщики с минимальным доходом до 30к (Группа Е) долгов больше других категорий 9,1% далее заемщики с уровнем догхода от 30к до 50к (группа D) имеют самое меньшее число долгов 6%, категория заемщиков с доходами от 50к до 200к  (группа С) на втором месте по количеству должников с уровнем в 8,5%, далее с уровнем раста доходов от 200к до 1000к (группа В) уровень должников снижается до 7,1%, и категория с доходами больше 1000к (группа А) имеют 8% должников.

**Дополнение к выводу выше:**  
В группы В и С входит 5014 и 15921 человек соответственно, поэтому будем считать эти выборки репрезентативными. Из количества должников в этих группах можно сделать вывод, что у заемщики с более высоким доходом чаще выплачивают долги.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [29]:
### Сгруппируем наши данные, что бы можно было подситать цель кредита влияет на количество должников.

purpose_debt = data.groupby('purpose_category').agg({'debt':['mean', 'sum', 'count']})

In [30]:
purpose_debt.round(3)

debt            
                           mean  sum  count
purpose_category                           
операции с автомобилем    0.093  400   4279
операции с недвижимостью  0.073  780  10751
получение образования     0.093  369   3988
проведение свадьбы        0.079  183   2313

**Вывод:** 
Из этих данных видно, что среди заемщиков, берущих кредит на операции с автомибелем и получение образования, наибольшее число должников с показателем 9,3%, возможно это связано с тем, что расходы на подобные цели тяжелее отбить быстро.
Заемщики, берущие кредит на проведение свадьбы, идут на третьем месте с показателем 7,9% должников, это может быть связано с тем, что на расходы на свадьбы меньше, а также часто на свадьбу дарят деньги, что помогает быстрее вернуть кредит.
И самые надежные заемщики банка берут кредит на операции с недвижимостью, количество должников среди них 7,3%, Это можно объяснить постоянным ростом цен на недвижимость, к тому же это один из самых надежных способов сбережения средств.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Количество данных с пропусками в стоблцах 'days_employed' и 'total_income' совпадает, а значит тут может быть несколько причин:
* Возможно в таблицу попали данные о людях не имеющих официального стажа, а значит и дохода. Поэтому эти данные не заполнены.
* Возможно в форме, которую заполняни в банке небыло необходимых пунков, которые корректно отражали бы ситуацию.

Пропуски могли появится в следующих случаях:
* Случайные пропуски - значения, для которых нет подходящей категории.
* Неслучайные пропуски - значения, для которых важен характер опроса или ответ, данный ранее (для другого столбца).

Есть еще полностью случайные пропуски, но эти возможные варианты я описал в предыдущем выводе.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
Как мы подсчитали выше пропуски составляли около 10% данных всего датафрейма, удалять их было бы некорректно, так как такое количество данных повлияло бы на результаты нашего исследования, поэтому мы заполнили данные наиболее близкими - медианными значениями доходов/величины стажа относительно каждой категории занятости заемщиков.

Мы использовали медиану для заполнеия пропусков, так как она наиболее точно позволяет заполнить значения опираясь на определенную группу. Медиана позволит избежать слышком завышенных или слишком заниженных значений, так как это такое значение, что половина выборки больше его, а половина выборки меньше его.

Среднее значене не подходит для нашего кейса, так как это сумма всех значений разделенное на количество значений, поэтому слишком высокие или слишком низкие значения данных могут повлиять на среднее, а значит и исказить заполняемые данные.

### Шаг 4: общий вывод.

Проект выполнялся по заданию кредитного отдела банка. Главный вопрос - влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок - да влияет. Эта зависимость прямая.

Опираяь только на те данные, котрые у нас есть можно сделать вывод, что идеальным заемщиков является человек без детей, состоящий в официальных отношениях, с уровнем доходов от 200к до 1000к (группа В), берущий кредит на операции с недвижимостью.

Худшим заемщиком является человек с 4-мя и более детьми, холостые с доходами от 50к до 200к (группа С), берущие займ на операции с автомобилем.

По остальным группам дохода сделать подобные выводы будет сложнее, так как эти группы заемщиков тяжело назвать репрезентативными.

Субъектом проекта является - статистика о платёжеспособности клиентов, с помощью которой мы ответили на следующие вопросы:

* Есть ли зависимость между количеством детей и возвратом кредита в срок - Да.
* Есть ли зависимость между семейным положением и возвратом кредита в срок - да.
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок - да.
* Как разные цели кредита влияют на его возврат в срок - Да.

Также мы составили портрет "лучшего" и "худшего" заемщиков.